### ISSUE TITLE
Unable to connect to Remedy API from XL Release

### ISSUE THREAD

**Issue by vishnukadi** [xebialabs-community/xlr-remedy-plugin#1]
Failed trying to connect to Remedy API from XL Release

Title: Remedy
Url: https://workspace-dev-restapi.onbmc.com/api
Authentication Method: Basic
Username: xxxxxx
Password: *******

Error details:
```
<html>
    <head>
        <meta http-equiv="Content-Type" content="text/html;charset=utf-8"/>
        <title>Error 500 Server Error</title>
    </head>
    <body>
        <h2>HTTP ERROR 500</h2>
        <p>Problem accessing /api/jwt/login. Reason:
 
            <pre>    Server Error</pre>
        </p>
        <h3>Caused by:</h3>
        <pre>java.lang.RuntimeException: java.text.ParseException: Invalid serialized unsecured/JWS/JWE object: Missing part delimiters
    at com.bmc.arsys.restapi.ARRequestFilter.readAuthorizationHeaderField(ARRequestFilter.java:108)
    at com.bmc.arsys.restapi.ARRequestFilter.doFilter(ARRequestFilter.java:65)
    at org.eclipse.equinox.http.servlet.internal.FilterRegistration.doFilter(FilterRegistration.java:81)
    at org.eclipse.equinox.http.servlet.internal.FilterChainImpl.doFilter(FilterChainImpl.java:35)
    at com.bmc.arsys.restapi.SSOWrapperFilter.doFilter(SSOWrapperFilter.java:49)
    at org.eclipse.equinox.http.servlet.internal.FilterRegistration.doFilter(FilterRegistration.java:81)
    at org.eclipse.equinox.http.servlet.internal.FilterChainImpl.doFilter(FilterChainImpl.java:35)
    at org.eclipse.equinox.http.servlet.internal.ProxyServlet.processAlias(ProxyServlet.java:132)
    at org.eclipse.equinox.http.servlet.internal.ProxyServlet.service(ProxyServlet.java:68)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
    at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:835)
    at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:583)
    at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:224)
    at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1158)
    at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:511)
    at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
    at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1090)
    at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
    at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:213)
    at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:109)
    at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:119)
    at org.eclipse.jetty.server.Server.handle(Server.java:517)
    at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:308)
    at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:242)
    at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:273)
    at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:95)
    at org.eclipse.jetty.io.SelectChannelEndPoint$2.run(SelectChannelEndPoint.java:75)
    at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceAndRun(ExecuteProduceConsume.java:213)
    at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:147)
    at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:654)
    at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:572)
    at java.lang.Thread.run(Thread.java:748)
Caused by: java.text.ParseException: Invalid serialized unsecured/JWS/JWE object: Missing part delimiters
    at com.nimbusds.jose.JOSEObject.split(JOSEObject.java:204)
    at com.nimbusds.jwt.SignedJWT.parse(SignedJWT.java:93)
    at com.bmc.arsys.restapi.ARRequestFilter.readAuthorizationHeaderField(ARRequestFilter.java:89)
    ... 31 more
</pre>
        <h3>Caused by:</h3>
        <pre>java.text.ParseException: Invalid serialized unsecured/JWS/JWE object: Missing part delimiters
    at com.nimbusds.jose.JOSEObject.split(JOSEObject.java:204)
    at com.nimbusds.jwt.SignedJWT.parse(SignedJWT.java:93)
    at com.bmc.arsys.restapi.ARRequestFilter.readAuthorizationHeaderField(ARRequestFilter.java:89)
    at com.bmc.arsys.restapi.ARRequestFilter.doFilter(ARRequestFilter.java:65)
    at org.eclipse.equinox.http.servlet.internal.FilterRegistration.doFilter(FilterRegistration.java:81)
    at org.eclipse.equinox.http.servlet.internal.FilterChainImpl.doFilter(FilterChainImpl.java:35)
    at com.bmc.arsys.restapi.SSOWrapperFilter.doFilter(SSOWrapperFilter.java:49)
    at org.eclipse.equinox.http.servlet.internal.FilterRegistration.doFilter(FilterRegistration.java:81)
    at org.eclipse.equinox.http.servlet.internal.FilterChainImpl.doFilter(FilterChainImpl.java:35)
    at org.eclipse.equinox.http.servlet.internal.ProxyServlet.processAlias(ProxyServlet.java:132)
    at org.eclipse.equinox.http.servlet.internal.ProxyServlet.service(ProxyServlet.java:68)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
    at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:835)
    at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:583)
    at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:224)
    at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1158)
    at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:511)
    at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
    at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1090)
    at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
    at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:213)
    at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:109)
    at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:119)
    at org.eclipse.jetty.server.Server.handle(Server.java:517)
    at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:308)
    at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:242)
    at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:273)
    at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:95)
    at org.eclipse.jetty.io.SelectChannelEndPoint$2.run(SelectChannelEndPoint.java:75)
    at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceAndRun(ExecuteProduceConsume.java:213)
    at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:147)
    at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:654)
    at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:572)
    at java.lang.Thread.run(Thread.java:748)
</pre>
    </body>
</html>
```

**Able to successfully login and get a response using Postman:**
Login call is made by passing key/value pairs for username and password in the body.
Content-Type is passed as application/x-www-form-urlencoded in header.
Token comes back as the response in Body, passed in authorization header by subsequent REST calls.

**Request 1**
URL: https://mastercard-dev-restapi.onbmc.com/api/jwt/login
Method: POST
Content-Type: application/x-www-form-urlencoded
Username/password sent as key value pairs in the body
**Response**
eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiIwXC8rMG9RbGs5VVlhZEUraVZMU1R6YUFxWExVWWxEbCtaSVFjbiswQkM2bTY3SjVSREJGdkkyU3hac1JpYTY0QTVNRHhOTUZNd3ZwMWExUnJaVG5rcjNUV1dFdHVEVUZmakxsQTFSTkRCUllTZ1VoV3VYdStDQT09IiwibmJmIjoxNTU2MjA4NTg1LCJpc3MiOiJtY3RsLWR2LWFybC0xIiwiZXhwIjoxNTU2MjEyMzA1LCJfY2FjaGVJZCI6MTYxMjEzLCJpYXQiOjE1NTYyMDg3MDUsImp0aSI6IklER0FBNVYwRjFLNUhBUFE4WTFMUFBMTFNXVFdXTSJ9.QHGTpkr5Wzu_0FGfvd12TuU68VvauLhgWaTzzU1oSac

**Request 2**
The authorization and content type are passed in the header 
URL: https://mastercard-dev-restapi.onbmc.com/api/arsys/v1/entry/CTM:People?q=%27Remedy%20Login%20ID%27%3d%22e025534%22
Method: GET
Content-Type: application/json
Authorization: eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiIwXC8rMG9RbGs5VVlhZEUraVZMU1R6YUFxWExVWWxEbCtaSVFjbiswQkM2bTY3SjVSREJGdkkyU3hac1JpYTY0QTVNRHhOTUZNd3ZwMWExUnJaVG5rcjNUV1dFdHVEVUZmakxsQTFSTkRCUllTZ1VoV3VYdStDQT09IiwibmJmIjoxNTU2MjA4NTg1LCJpc3MiOiJtY3RsLWR2LWFybC0xIiwiZXhwIjoxNTU2MjEyMzA1LCJfY2FjaGVJZCI6MTYxMjEzLCJpYXQiOjE1NTYyMDg3MDUsImp0aSI6IklER0FBNVYwRjFLNUhBUFE4WTFMUFBMTFNXVFdXTSJ9.QHGTpkr5Wzu_0FGfvd12TuU68VvauLhgWaTzzU1oSac
**Response**
Response comes back successfully as response.entries[ ].values[ ]
```

{
    "entries": [
     ..................   JSON response  ..................
                    }	

```

**Comment by vlussenburg** [comment]
HI @vishnukadi. Could you try configuring in the Shared Configuration without the '/api' in the URL field?

Url: https://workspace-dev-restapi.onbmc.com/api

should be 

Url: https://workspace-dev-restapi.onbmc.com

**Comment by vishnukadi** [comment]
Hi @vlussenburg 
Tried changing the url without '/api', but received same error.

**Comment by vlussenburg** [comment]
Thanks for trying this out. 

According to the [REST API documentation](https://docs.bmc.com/docs/ars1902/authentication-and-permissions-in-the-rest-api-847208964.html) the value of the Authorization header should be prefixed with `AR-JWT ` which I don't see in your Postman description. What version of Remedy are you working with?

**Comment by vishnukadi** [comment]
Remedy version: 9.1

I just tried again with the authorization header prefixed with AR-JWT and I received a successful response.

Authorization header : AR-JWT {{jwt}}
token stored in a postman variable

**Comment by vlussenburg** [comment]
Hi @vishnukadi ! So it worked both with `AR-JWT` and without it? Interesting! 

The error you're getting suggest ARS is trying to read a Authorization header before the login is complete, which of course doesn't make sense when doing the login call, the Authorization header shouldn't be read there. So I'm not sure what is going on!

What you could try is to [enable the debug logging for http components](https://support.xebialabs.com/hc/en-us/community/posts/204756423-How-To-Configure-additional-logging-for-Apache-HttpComponents-i-e-HttpRequest-and-HttpResponse). Maybe we'll get some clues from that.

**Comment by vishnukadi** [comment]
Hi @vlussenburg 

I enabled debug logging and did a test connection from the plugin. 

Here is the log output
```
2019-05-02 14:04:41.658 DEBUG [o.a.h.c.s.SSLConnectionSocketFactory] Starting handshake
2019-05-02 14:04:41.753 DEBUG [o.a.h.c.s.SSLConnectionSocketFactory] Secure session established
2019-05-02 14:04:41.753 DEBUG [o.a.h.c.s.SSLConnectionSocketFactory]  negotiated protocol: TLSv1.2
2019-05-02 14:04:41.753 DEBUG [o.a.h.c.s.SSLConnectionSocketFactory]  negotiated cipher suite: TLS_RSA_WITH_AES_256_CBC_SHA256
2019-05-02 14:04:41.753 DEBUG [o.a.h.c.s.SSLConnectionSocketFactory]  peer principal: CN=*.onbmc.com, O=BMC Software Inc, L=Houston, ST=Texas, C=US
2019-05-02 14:04:41.753 DEBUG [o.a.h.c.s.SSLConnectionSocketFactory]  peer alternative names: [*.onbmc.com, onbmc.com]
2019-05-02 14:04:41.753 DEBUG [o.a.h.c.s.SSLConnectionSocketFactory]  issuer principal: CN=DigiCert SHA2 Secure Server CA, O=DigiCert Inc, C=US
2019-05-02 14:04:41.754 DEBUG [o.a.h.i.c.DefaultHttpClientConnectionOperator] Connection established 10.157.147.3:40342<->216.52.6.177:443
2019-05-02 14:04:41.754 DEBUG [o.a.h.impl.execchain.MainClientExec] Executing request POST /api/jwt/login HTTP/1.1
2019-05-02 14:04:41.754 DEBUG [o.a.h.impl.execchain.MainClientExec] Proxy auth state: UNCHALLENGED
2019-05-02 14:04:41.754 DEBUG [org.apache.http.headers] http-outgoing-9 >> POST /api/jwt/login HTTP/1.1
2019-05-02 14:04:41.754 DEBUG [org.apache.http.headers] http-outgoing-9 >> Content-Type: application/x-www-form-urlencoded
2019-05-02 14:04:41.754 DEBUG [org.apache.http.headers] http-outgoing-9 >> Accept: text/plain
2019-05-02 14:04:41.754 DEBUG [org.apache.http.headers] http-outgoing-9 >> Authorization: Basic <<HIDDEN>>
2019-05-02 14:04:41.754 DEBUG [org.apache.http.headers] http-outgoing-9 >> Content-Length: 36
2019-05-02 14:04:41.754 DEBUG [org.apache.http.headers] http-outgoing-9 >> Host: workspace-dev-restapi.onbmc.com
2019-05-02 14:04:41.754 DEBUG [org.apache.http.headers] http-outgoing-9 >> Connection: Keep-Alive
2019-05-02 14:04:41.754 DEBUG [org.apache.http.headers] http-outgoing-9 >> User-Agent: Apache-HttpClient/4.5.3 (Java/1.8.0_191)
2019-05-02 14:04:41.754 DEBUG [org.apache.http.headers] http-outgoing-9 >> Accept-Encoding: gzip,deflate
2019-05-02 14:04:41.815 DEBUG [org.apache.http.headers] http-outgoing-9 << HTTP/1.1 500 Server Error
2019-05-02 14:04:41.815 DEBUG [org.apache.http.headers] http-outgoing-9 << Date: Thu, 02 May 2019 19:04:41 GMT
2019-05-02 14:04:41.815 DEBUG [org.apache.http.headers] http-outgoing-9 << X-Frame-Options: SAMEORIGIN
2019-05-02 14:04:41.815 DEBUG [org.apache.http.headers] http-outgoing-9 << Content-Security-Policy: frame-ancestors 'self'
2019-05-02 14:04:41.815 DEBUG [org.apache.http.headers] http-outgoing-9 << Content-Type: text/html;charset=iso-8859-1
2019-05-02 14:04:41.815 DEBUG [org.apache.http.headers] http-outgoing-9 << Cache-Control: must-revalidate,no-cache,no-store
2019-05-02 14:04:41.815 DEBUG [org.apache.http.headers] http-outgoing-9 << Content-Length: 1122
2019-05-02 14:04:41.815 DEBUG [org.apache.http.headers] http-outgoing-9 << Connection: close
2019-05-02 14:04:41.815 DEBUG [org.apache.http.headers] http-outgoing-9 << Set-Cookie: onbmc_pool=!9YaBz7mYdmYsaK3hXnhJ38sCg2bkJz2Ct80+MZ9tOk3NJJigIGi29Gz75dc75S+RQxZFNRbgdSarzt8=; path=/; Httponly
2019-05-02 14:04:41.815 DEBUG [org.apache.http.headers] http-outgoing-9 << Vary: Accept-Encoding
2019-05-02 14:04:41.815 DEBUG [org.apache.http.headers] http-outgoing-9 << Content-Encoding: gzip
2019-05-02 14:04:41.816 DEBUG [o.a.h.c.p.ResponseProcessCookies] Cookie accepted [onbmc_pool="!9YaBz7mYdmYsaK3hXnhJ38sCg2bkJz2Ct80+MZ9tOk3NJJigIGi29Gz75dc75S+RQxZFNRbgdSarzt8=", version:0, domain:workspace-dev-restapi.onbmc.com, path:/, expiry:null]
2019-05-02 14:04:41.817 DEBUG [o.a.h.i.c.DefaultManagedHttpClientConnection] http-outgoing-9: Close connection
2019-05-02 14:04:41.818 DEBUG [o.a.h.impl.execchain.MainClientExec] Connection discarded
2019-05-02 14:04:41.818 DEBUG [o.a.h.i.c.PoolingHttpClientConnectionManager] Connection released: [id: 9][route: {s}->https://workspace-dev-restapi.onbmc.com:443][total kept alive: 0; route allocated: 0 of 2; total allocated: 0 of 20]
2019-05-02 14:04:41.819 DEBUG [o.a.h.i.c.PoolingHttpClientConnectionManager] Connection manager is shutting down
2019-05-02 14:04:41.819 DEBUG [o.a.h.i.c.PoolingHttpClientConnectionManager] Connection manager shut down
```


It looks like the header `Authorization: Basic <<HIDDEN>>` has to be disabled in the login call. 

Authorization type is set to 'no auth' in postman for the successful test I did earlier. And when I change the Authorization to "Basic" in postman, I get a 500 error response.

I checked this using curl command as well, there is no authorization header being passed - 

```
curl -v -d "username=username&password=*******" -X POST https://workspace-dev-restapi.onbmc.com/api/jwt/login


* About to connect() to workspace-dev-restapi.onbmc.com port 443 (#0)
*   Trying 216.52.6.177... connected
* Connected to workspace-dev-restapi.onbmc.com (216.52.6.177) port 443 (#0)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
*   CAfile: /etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem
  CApath: none
* SSL connection using TLS_RSA_WITH_AES_256_CBC_SHA256
* Server certificate:
*       subject: CN=*.onbmc.com,O=BMC Software Inc,L=Houston,ST=Texas,C=US
*       start date: Mar 06 00:00:00 2017 GMT
*       expire date: Jun 03 12:00:00 2020 GMT
*       common name: *.onbmc.com
*       issuer: CN=DigiCert SHA2 Secure Server CA,O=DigiCert Inc,C=US
> POST /api/jwt/login HTTP/1.1
> User-Agent: curl/7.19.7 (x86_64-redhat-linux-gnu) libcurl/7.19.7 NSS/3.36 zlib/1.2.3 libidn/1.18 libssh2/1.4.2
> Host: workspace-dev-restapi.onbmc.com
> Accept: */*
> Content-Length: 36
> Content-Type: application/x-www-form-urlencoded
>
< HTTP/1.1 200 OK
< Date: Thu, 02 May 2019 19:28:41 GMT
< X-Frame-Options: SAMEORIGIN
< Content-Security-Policy: frame-ancestors 'self'
< Content-Type: text/plain
< Content-Length: 413
< Set-Cookie: onbmc_pool=!VMJzr/44gL42x/LhXnhJ38sCg2bkJwshGVKNU3n7h/hf9c7Qi7djxwDgdRs9or7BwOvuy2AKK5YlmOU=; path=/; Httponly
<
* Connection #0 to host workspace-dev-restapi.onbmc.com left intact
* Closing connection #0
eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJESVZXOXNKSVA5QVFYeUdTcVhnZG9TNWNqTUFcL1Nja05XeStpdkI3cm5TWkJtdGNDdFd5UXBSbGRPbUJ3R01sMk90RThaOHpRcVJKMDhhS01pUHNUN01IMGxFMmdDZHFlMlE2dE55ajBseEhNNlVxRG5YRWozUT09IiwibmJmIjoxNTU2ODI1MjAxLCJpc3MiOiJtY3RsLXBkLWFybC0yIiwiZXhwIjoxNTU2ODI1NjgxLCJfY2FjaGVJZCI6MTg0MjYzLCJpYXQiOjE1NTY4MjUzMjEsImp0aSI6IklER0FBNVYwRjE5Nk1BUFFXR0RUUFBZVENLM05DTyJ9.MDUm9sO-7pSDXOQ14fQPhwaGQeCojYPyfstV6oWw7yQ
```

**Comment by vlussenburg** [comment]
I think I know what's going on here. The standard 'basic authentication' implementation is kicking in. Could you change the dropdown from 'Basic' to blank (in the Shared Configuration menu in XL Release)?

**Comment by vishnukadi** [comment]
I tried changing the dropdown to blank, but still basic authentication header is being passed. 
Tried changing it to "None' as well, still seeing the same error.

**Comment by vlussenburg** [comment]
That's odd! What version of XL Release are you on? It should only apply the credentials when it's set to 'Basic'.

Regardless, I'll update the implementation so it never passes the credentials to the underlying libraries.

**Comment by vlussenburg** [comment]
I'm testing the implementation, if you want to test it on your end:

- Save the contents of [this file](https://github.com/xebialabs-community/xlr-remedy-plugin/blob/master/src/main/resources/remedy/RemedyClient.py) to `XL_RELEASE_HOME/ext/remedy/RemedyClient.py`.
- Restart XL Release
- Test :)

**Comment by vlussenburg** [comment]
LGTM - released [v3.0.0.1](https://github.com/xebialabs-community/xlr-remedy-plugin/releases/download/v3.0.0.1/xlr-remedy-plugin-3.0.0.1.jar)

**Comment by vishnukadi** [comment]
I am able to connect successfully now. 
Thanks a ton @vlussenburg for your help.
